src/createGlobalState.ts

```diff
@@ -1,19 +1,38 @@
 /* eslint-disable */
-import { useLayoutEffect, useState } from 'react';
+import { useLayoutEffect, useState, Dispatch } from 'react';
+import { resolveHookState, InitialHookState, HookState } from './util/resolveHookState'
 import useEffectOnce from './useEffectOnce';
 
-export function createGlobalState<S = any>(initialState?: S) {
-  const store: { state: S | undefined; setState: (state: S) => void; setters: any[] } = {
-    state: initialState,
-    setState(state: S) {
-      store.state = state;
+export type GlobalStateHookReturn<S> = [ S, Dispatch<HookState<S>> ]
+export type GlobalStateHook<S> = (initialState?: InitialHookState<S>) => GlobalStateHookReturn<S>
+interface StateStore<S> {
+  initialized: boolean
+  state: S
+  setState: Dispatch<HookState<S>>
+  setters: Dispatch<HookState<S>>[]
+}
+export function createGlobalState<S = any>(defaultState: S): GlobalStateHook<S> {
+  const store: StateStore<S> = {
+    initialized: false,
+    state: defaultState,
+    setState(newState: HookState<S>) {
+      store.state = resolveHookState(newState);
       store.setters.forEach(setter => setter(store.state));
     },
     setters: [],
   };
 
-  return (): [S | undefined, (state: S) => void] => {
-    const [globalState, stateSetter] = useState<S | undefined>(store.state);
+  // unlike in useState, here initialState is optional because there will always
```

Comment:
`useState` does not require an initial state if you accept `undefined` as a possible value.

Reply:
I think my comment wasn't quite clear. I'm contrasting `useState` with the `useGlobalState` created by this function. For example, if you have a state
```ts
const [myDefinedState, setMyDefinedState] = useState<string>(INITIAL_STRING)
```
`INITIAL_STRING` is required, because the type does not include `undefined` in its definition.
In contrast:
```ts
const useMyDefinedState = createGlobalState<string>(DEFAULT_STRING)
....
const [ myDefinedState, setMyDefinedState ] = useMyDefinedState()
```
The initializer for `useMyDefinedState` is not required because it was previously supplied by the default param passed to `createGlobalState`.

Reply:
React includes two definitions of `useState`. The one without a parameter automatically adds `undefined` to the value type:

```ts
function useState<S = undefined>(): [S | undefined, Dispatch<SetStateAction<S | undefined>>];
```

So if you have `const [myDefinedState, setMyDefinedState] = useState<string>()`, the type of `myDefinedState` becomes `string | undefined`.

Of course, this is still different from your suggested `useGlobalState`, which doesn't change the value type.

Reply:
Oh, I didn't realize react did that. In that case, it sounds like just removing that comment will address your review?

Reply:
Yes, please fix the comment.

Reply:
Fixed